process_blocking.go

```diff
@@ -317,6 +319,35 @@ func (p *blockingProcess) Wait(ctx context.Context) error {
 	}
 }
 
+func (p *blockingProcess) Restart(ctx context.Context) error {
+	if !p.Complete(ctx) {
+		if err := Terminate(ctx, p); err != nil {
+			return err
+		}
+		// This call will almost always fail, since the process was terminated with
+		// a signal and not allowed to exit successsfully. Therefore, we don't check
+		// for error here.
+		p.Wait(ctx)
+	}
+	// By now, this process should be dead.
+
+	p.info = nil
+	p.opts.closers = []func(){}
+
+	cmd, err := p.opts.Resolve(ctx)
+	if err != nil {
+		return errors.Wrap(err, "problem building command from options")
+	}
+
+	if err = cmd.Start(); err != nil {
+		return errors.Wrap(err, "problem restarting command")
+	}
+
+	go p.reactor(ctx, cmd)
```

Comment:
We _could_ use the constructor and that is what I wanted to do at first, since something like this looks a lot cleaner and re-uses code:
```go
func (p *blockingProcess) Restart(ctx context.Context) error {
    opts := p.Info(ctx).Options
    proc, err := newBlockingProcess(ctx, &opts)

    p = proc.(*blockingProcess)
    return err
}
```

However, I went against this because `newBlockingConstructor()` creates a complete new and separate process with a new ID. This means that when a user runs `p.Restart()`, they won't actually be able to restart their original process, they are changing `p` to become a separate and new process. This seems like a side-effect that I could not reasonably include in a method called `Restart()` since it implies that everything is the same but the process is simply restarted.

Reply:
oh, I always thought that you would want a new process ID, because otherwise things associated with the previous process (e.g. it's outcome, in memory logs, etc would get lost or overwritten. Maybe the right signature for this is something that returns a Process rather than just an error? 

We could make it be called "Respawn" for clarity.

Reply:
That sounds fine to me then. A rename + change to using the constructor resolves the issue with clarity. I'll edit this message when I've pushed the required changes to use a different process instead of re-running the current one.